src/table/boot.rs

```diff
@@ -993,6 +1000,41 @@ impl BootServices {
             OpenProtocolAttributes::Exclusive,
         )
     }
+
+    /// Get a NULL-terminated Path string the given image was loaded from.
+    ///
+    /// Note the path string will be empty if the image was loaded from a
+    /// buffer instead of a path, and the path wasn't set.
+    pub fn get_image_file_path(&self, image_handle: Handle) -> Result<&CStr16> {
+        let loaded_image = self.open_protocol::<LoadedImage>(
+            OpenProtocolParams {
+                handle: image_handle,
+                agent: image_handle,
+                controller: None,
+            },
+            OpenProtocolAttributes::Exclusive,
+        )?;
+        let loaded_image = unsafe { &*loaded_image.interface.get() };
+
+        // If the file_path is None, it means the operation isn't supported.
+        let file_path = loaded_image.file_path().ok_or(Status::UNSUPPORTED)?;
+
+        // check if the type of `file_path` is correct
+        if let (DeviceType::MEDIA, DeviceSubType::MEDIA_FILE_PATH) =
+            (file_path.device_type(), file_path.sub_type())
+        {
+            let file_path = file_path as *const DevicePath;
+
+            unsafe {
+                // path name follows by the `DevicePathHeader`
+                let path_name = file_path.offset(1).cast();
+                Ok(CStr16::from_ptr(path_name))
```

Comment:
I think that this conversion can end up being undefined behavior because the path string is not guaranteed to be properly aligned. I've put up https://github.com/rust-osdev/uefi-rs/pull/404 to implement a new `FilePathMediaDevicePath` type, along with an `UnalignedCStr16` which will help handle this safely.

So, what I'd like to propose here is to drop `get_image_file_path` from this PR, and then once both this PR and the one linked above are merged you'll be able to do something like this:
```rust
let path = loaded_image
  .file_path()?
  .as_file_path_media_device_path()?
  .path_name()
  .to_cstring16().ok()?;
```

A bit more verbose, but I think it will be a good safe and composable API. Does that sound reasonable to you?

Reply:
> I think that this conversion can end up being undefined behavior because the path string is not guaranteed to be properly aligned. I've put up #404 to implement a new `FilePathMediaDevicePath` type, along with an `UnalignedCStr16` which will help handle this safely.
> 
> So, what I'd like to propose here is to drop `get_image_file_path` from this PR, and then once both this PR and the one linked above are merged you'll be able to do something like this:
> 
> ```rust
> let path = loaded_image
>   .file_path()?
>   .as_file_path_media_device_path()?
>   .path_name()
>   .to_cstring16().ok()?;
> ```
> 
> A bit more verbose, but I think it will be a good safe and composable API. Does that sound reasonable to you?

That's good, I drop `get_image_file_path` now.😃